Split aria-state-or-property-permitted-5c01ea.md (#2036)
auto-wcag/auto-wcag: _rules/aria-state-or-property-not-prohibited-yyw7k3.md

accessibility_requirements:
  wcag-technique:ARIA5: # Using WAI-ARIA state and property attributes to expose the state of a user interface component
    forConformance: false
    failed: not satisfied
    passed: further testing needed
    inapplicable: further testing needed
  aria12:state_property_processing:
    title: ARIA 1.2, 8.6 State and Property Attribute Processing
    forConformance: true
    failed: not satisfied
    passed: satisfied
    inapplicable: satisfied
  wcag20:1.3.1: # Info and Relationships (A)
    secondary: true
  wcag20:4.1.2: # Name, Role, Value (A)
    secondary: true
input_aspects:
  - Accessibility Tree
  - CSS styling
  - DOM Tree
acknowledgments:
  authors:
    - Trevor Bostic
  previous_authors:
    - Anne Thyme Nørregaard
    - Jean-Yves Moyen
  funding:
    - WAI-Tools
  assets:
    - JFK's "We Choose the Moon" speech excerpt is courtesy of NASA.
---

## Applicability

This rule applies to any [WAI-ARIA state or property][] that is specified on an [HTML or SVG element][namespaced element] that is [included in the accessibility tree][].

## Expectation

> No test target (i.e., state or property) is prohibited on the semantic role of the element on which it is specified.

Reviewer: I think it would be good for the word "prohibited" to link to the corresponding section in WAI-ARIA. There are lots of ways in which an attribute may not be allowed, and without a link this could be misunderstood to be broader than it actually is.

> No test target (i.e., state or property) is prohibited on the semantic role of the element on which it is specified.

Reviewer: I'm not a fan of the "(i.e. state or property)" bit. If we were to do this in all rules I'm concerned this would get messy. If we feel that "test target" as a term is confusing we should have a separate conversation about that, and then adjust all rules to fit the outcome. I don't think we want to introduce changes like this without thinking through the broader implications.

Assessing the value of the attribute is out of scope for this rule.

### Related rules

- [ARIA state or property has valid value](https://www.w3.org/WAI/standards-guidelines/act/rules/6a7281/proposed/)
- [ARIA state or property is permitted](https://www.w3.org/WAI/standards-guidelines/act/rules/5c01ea/proposed/)

### Bibliography

- [Understanding Success Criterion 4.1.1: Parsing](https://www.w3.org/WAI/WCAG21/Understanding/parsing.html)
- [Understanding Success Criterion 4.1.2: Name, Role, Value](https://www.w3.org/WAI/WCAG21/Understanding/name-role-value.html)
- [WAI-ARIA 1.2, Supported States and Properties](https://www.w3.org/TR/wai-aria-1.2/#states_and_properties)
- [WAI-ARIA 1.2, Global States and Properties](https://www.w3.org/TR/wai-aria-1.2/#global_states)
- [ARIA5: Using WAI-ARIA state and property attributes to expose the state of a user interface component](https://www.w3.org/WAI/WCAG21/Techniques/aria/ARIA5)
- [Document conformance requirements for use of ARIA attributes in HTML](https://www.w3.org/TR/html-aria/#docconformance)

## Test Cases

> ### Passed

Reviewer: `aria-pressed` is never prohibited. I think it would be good to (also?) include examples of attributes that can be prohibited.

---
id: yyw7k3

> name: ARIA state or property is not prohibited

Reviewer: How would you feel about having this rule as a replacement of "Element marked as decorative is not exposed"? I think right now this rule avoids most overlap between the two rules, but it seems to me that the way the ARIA WG intends prohibited attributes to work, this rule isn't broad enough, and most of what it's missing is covered by the other rule. I left a couple suggestions, if you want to go that direction, I think we should consider deprecating the other rule.

@Jym77 I would appreciate it you'd weigh in on this too.

Author: Sorry, I apparently missed this comment on my first look through. I agree with the sentiment here though, if we want to expand this rule with the applicability you suggested, then I think we should deprecate "Element marked as decorative is not exposed". Additionally, since "Element marked as decorative" doesn't have any accessibility requirements, including those failed cases under this rule would give more direction on what we think reporting requirements are.

TL;DR - In favor of expanding this rule and deprecating "Element marked as decorative"

Reviewer 2: Seemed to have missed the initial comment also… No fully sure I remember all the details, but…

* I believe that "element marked as decorative" (together with "`aria-hidden` no focusable content" and "presentational children no focusable content") can be consolidated into "presentational conflict is not triggered" (#1507). It seems all 3 check a different aspect of the conflict. I've also gathered from a couple of discussions that ARIA group more or less considers triggering the conflict as an author error (even though it didn't make to the spec).
* This rule is conceptually a bit different since prohibited attributes could be something else. Notably, some roles have prohibited naming (and thus attributes `aria-label(led-by)`) without triggering the conflict.
* I somewhat remember a discussion (maybe here 😅) about the fact that `role="none" aria-label="foo"` does trigger the conflict, thus the role is not `none` and the element is included in the accessibility tree (plus, the `aria-label` may be not prohibited on the effective role). I think we concluded that ARIA intention was to consider this as a prohibited attribute nonetheless.

So, I'm rather in favour of deprecating "element marked as decorative" which is a bit half-baked and too narrow in its focus, and that we can do much better in terms of intention now. But I'm not sure that this rule is the best fit for the deprecation. I'm also not against a bit of overlap between "attribute not prohibited" and "conflict not triggered" rules, given that they do test with very different focus in mind, and also report a lot of different things on their own.

rule_type: atomic
description: |
  This rule checks that WAI-ARIA states or properties are not prohibited for the element they are specified on.
accessibility_requirements:
  wcag-technique:ARIA5: # Using WAI-ARIA state and property attributes to expose the state of a user interface component
    forConformance: false
    failed: not satisfied
    passed: further testing needed
    inapplicable: further testing needed
  aria12:state_property_processing:
    title: ARIA 1.2, 8.6 State and Property Attribute Processing
    forConformance: true
    failed: not satisfied
    passed: satisfied
    inapplicable: satisfied
  wcag20:1.3.1: # Info and Relationships (A)
    secondary: true
  wcag20:4.1.2: # Name, Role, Value (A)
    secondary: true
input_aspects:
  - Accessibility Tree
  - CSS styling
  - DOM Tree
acknowledgments:
  authors:
    - Trevor Bostic
  previous_authors:
    - Anne Thyme Nørregaard
    - Jean-Yves Moyen
  funding:
    - WAI-Tools
  assets:
    - JFK's "We Choose the Moon" speech excerpt is courtesy of NASA.
---

## Applicability

> This rule applies to any [WAI-ARIA state or property][] that is specified on an [HTML or SVG element][namespaced element] that is [included in the accessibility tree][].

Reviewer: I don't know if you meant for this rule to skip prohibited attributes on role=none/presentation/generic, but by using Included in the accessibility tree that's what happens. Those aren't in the accessibility tree. That stuff is covered by Decorative element not exposed though, so maybe that's okay? I don't think ARIA WG intended for those to be ignored. In ARIA 1.2 that includes almost all prohibited attributes.

One thing you could do here is to use programmatically hidden instead:

```suggestion
This rule applies to any [WAI-ARIA state or property][] that is specified on an [HTML or SVG element][namespaced element] that is not [programmatically hidden][].
```

Author: Failed example 1 shows a failure on a generic role, so we would either need to change the applicability here, or we need to remove that example.

## Expectation

> No test target is [prohibited](https://www.w3.org/TR/wai-aria-1.2/#prohibitedattributes) on the semantic role of the element on which it is specified.

Reviewer: I wonder if semantic role is right here. It creates this strange interaction with presentational conflict resolution. If you agree with my suggestion on the applicability, the perhaps what needs to go here is something like this:

```suggestion
No test target is [prohibited](https://www.w3.org/TR/wai-aria-1.2/#prohibitedattributes) on the element's role. The role here is the [explicit role][] if one exists, or the [implicit role][] otherwise.

**note** The role here is different from the [semantic role][], as it does not [consider presentational role conflict resolution][].
```

## Assumptions

There are no assumptions.

## Accessibility Support

Implementation of [Presentational Roles Conflict Resolution][] varies from one browser or assistive technology to another. Depending on this, some elements can have a [semantic role][] of `none` and their attributes fail this rule with some technologies but users of other technology would not experience any accessibility issue.

## Background

The presence of prohibited ARIA attributes is often the result of a developer using an incorrect role, or a misunderstanding of the attribute. These attributes are ignored by browsers and other assistive technologies. This often means that a state or property which should exist is missing. This can cause issues under success criterion [1.3.1 Info and Relationships][sc131] or [4.1.2 Name, Rule Value][sc412].

Assessing the value of the attribute is out of scope for this rule.

### Related rules

> 

Reviewer: This rule has an interesting interaction with Decorative element not exposed. It almost seems to me that this rule should cover everything that rule does, except that you've sidestepped all the overlapping cases by looking at elements included in the accessibility tree, which avoid anything role=none/presentation.

```suggestion

- [Element marked as decorative is not exposed](https://www.w3.org/WAI/standards-guidelines/act/rules/46ca7f/)
```

Author: With your suggested change to the applicability, we would have some overlapping cases, which I think might just complicate things. As I've been digging more into it, I'm becoming more inclined to leave the applicability and expectation as they are. 

With that said, I think its still work adding this as a related rule so I'm going to add this suggestion in.

- [Document conformance requirements for use of ARIA attributes in HTML](https://www.w3.org/TR/html-aria/#docconformance)

## Test Cases

### Passed

#### Passed Example 1

The `aria-pressed` [state][] is [supported][] for the `button` role, which is the [implicit role][] for `button` elements.

```html
<button aria-pressed="false">My button</button>
```

#### Passed Example 2

The `aria-pressed` [state][] is [supported][] for the `button` role, which is the [explicit role][] of this `div` element.

```html
<div role="button" aria-pressed="false">My button</div>
```

#### Passed Example 3

The `aria-labelledby` [property][] is [supported][] for the `textbox` role, which is the [implicit role][] for and `input` element with `type="text"`.

```html
<input name="searchtxt" type="text" aria-labelledby="searchbtn" />
<input name="searchbtn" id="searchbtn" type="submit" value="Search" />
```

### Failed

#### Failed Example 1

The `aria-label` property is [prohibited](https://www.w3.org/TR/wai-aria-1.2/#prohibitedattributes) for an element with a generic role.

```html

> <div aria-label="Bananas"></div>

Reviewer: I'm not sure this element is actually in the accessibility tree. ARIA 1.3 draft contradicts itself on that:

> https://w3c.github.io/aria/#scope
> user agents MUST ignore any elements with the role generic or none.

> https://w3c.github.io/aria/#generic
> However, unlike elements with role presentation, generic elements are exposed in accessibility APIs

Author: I could leave or delete this example, I think we could point to this sentence and say that it should be in the tree. 

> However, unlike elements with role presentation, user agents expose generic elements in [accessibility APIs](https://w3c.github.io/aria/#dfn-accessibility-api) when permitted accessibility attributes have been specified.
